[WIP] [BUGFIX canary] Allow contextual components to be a glimmer component

Serabe opened the pull request:
Currently, contextual components does not work when the component being closured is a glimmer component. This PR aims to fix #12590 
- [x] ~~Add `auto` flag to `mut`s added by the compiler. (Needs some test)~~
- [x] ~~Make GlimmerComponents render as a contextual component~~
- [x] Check which tests would make sense to have in both components and Glimmer components.
- [x] ~~_Unbox_ auto muts when rendering a component using the component helper if the component is a GlimmerComponent~~
- [x] Make appropiate changes to the node manager.
- [x] ~~Test unboxing of `auto` `mut`s~~
- [x] Make `<my.comp>` work
- [x] Remove unboxing of `auto` `mut`s
- [x] Clean the code
- [x] Squash commits

Serabe committed:
[BUGFIX canary] Implement dot syntax for contextual Glimmer components

Example:

```hbs
<my-form model=model as |f|>
  <f.button name='firstName' />
  <f.submit>Save</f.submit>
</my-form>
```

rwjblue commented:
The path/checklist seems correct to me.

Serabe commented:
Quick update: I haven't forgot about this. I'm doing some tests on Glimmer components before going full speed into solving the issue. I have divide first task in two since I need to implement the other feature to test it.

Serabe commented:
`GlimmmerComponent`s can already be rendered as contextual components. I added the task of checking which tests in closure components would make sense to have for `GlimmerComponent`s as well. I'll be doing that next (read it as tomorrow).

I have yet to do the `mut` unboxing, but I think there is a simple way by adding an optional parameter to `getCellOrValue`. I should not be modifying any part of the render process any more (unless not significantly).

mixonic commented on packages/ember-htmlbars/lib/node-managers/component-node-manager.js:
```diff
@@ -83,7 +85,7 @@ ComponentNodeManager.create = function ComponentNodeManager_create(renderNode, e
       assert(`You cannot invoke the '${tagName}' component with angle brackets, because it's a subclass of Component. Please upgrade to GlimmerComponent. Alternatively, you can invoke as '{{${tagName}}}'.`, component.isGlimmerComponent);
     } else {
-      assert(`You cannot invoke the '${tagName}' component with curly braces, because it's a subclass of GlimmerComponent. Please invoke it as '<${tagName}>' instead.`, !component.isGlimmerComponent);
+      assert(`You cannot invoke the '${tagName}' component with curly braces, because it's a subclass of GlimmerComponent. Please invoke it as '<${tagName}>' instead.`, !component.isGlimmerComponent || isClosureComponent || isComponentHelper);
```
Is `isClosureComponent` threaded through here just for the assertion?

Perhaps we can just move the assertion further up the stack and avoid this. It kind of stinks to be passing these flags around.

Serabe replied:
Yes. Currently looking if I can just change `isAngleBracket` to mean _behave like a `GlimmerComponent`_. Otherwise, this can be hard to maintain.

Serabe commented:
Some update on this:

I removed `{auto: true}` from the transformation, given that it is already being marked as `internal`. Furthermore, I changed slightly the code from this:

```js
if (stream[MUTABLE_REFERENCE]) {
  return stream;
}
```

to:

```js
if (stream[MUTABLE_REFERENCE] && !internal) {
  return stream;
}
```

given that a mutable stream needs to be marked internal to be unboxed if needed.

Some changes were introduced in the rendering process too. Some conditionals depend on the component being a GlimmerComponent, but others actually depend on it being called with angle brackets. Would it be possible to talk to someone in Slack to check a few things?

Last, I'll spend some time adding more tests, specially from #12613 to see we support GlimmerComponents properly. After all is working, I'll clean the code.

Sorry for the delay. Still working on this.

mixonic commented:
@Serabe there has been some discussion about this lately, and it will need adjustment (sorry!). @wycats would like to hold off on invoking glimmer components with `{{`, which means they should not be invokable with the `{{component` helper or with `{{dot.syntax`.

However, dot syntax invocation with `<` should be supported. For example:

```hbs
<my-form model=model as |f|>
  <f.button name='firstName' />
  <f.submit>Save</f.submit>
</my-form>
```

Of course this only applies to glimmer components. Happy to chat via Slack this week.

homu commented:
:umbrella: The latest upstream changes (presumably #12682) made this pull request unmergeable. Please resolve the merge conflicts.

Serabe commented:
@homu, thank you. I'll be rebasing master soon, though I'll try to squash the commits so far (the conflict happens in `closure_component_test.js` and I hope that that is only because of 3477995, that gets mostly reverted in 9a9eab5)

Serabe commented:
This PR needs tildeio/htmlbars#438 due to `<my.component>` being compiled to the fragment part.

rwjblue commented:
@Serabe - Can you updated `package.json` and `npm-shrinkwrap.json` to use `htmlbars@0.14.10` (manually apply the same changes as https://github.com/emberjs/ember.js/pull/12705 for the version you need).

Serabe commented:
Yes. I'll upload a PR tomorrow morning for this. I'm not uploading this tonight cause npm is taking longer than usual and I cannot check the tests :disappointed:

Serabe commented:
@rwjblue I have uploaded PR #12706. I cannot install the dependencies using the npm shrinkwrap file, though I can install them without it. Cannot check in a different network until Monday.

rwjblue commented:
Thanks @serabe!

Serabe commented:
Adding some tests and then I'll remove the autounboxing feature. With the new requirements it is not needed.

homu commented:
:umbrella: The latest upstream changes (presumably #12712) made this pull request unmergeable. Please resolve the merge conflicts.

Serabe commented:
Rebased!

homu commented:
:umbrella: The latest upstream changes (presumably #12746) made this pull request unmergeable. Please resolve the merge conflicts.

homu commented:
:umbrella: The latest upstream changes (presumably #13096) made this pull request unmergeable. Please resolve the merge conflicts.

rwjblue commented:
I'm not sure of the path forward on this.  We should likely let the dust settle on the glimmer integration efforts and attack again....

rwjblue commented:
Thanks so much for working on this @Serabe, but since we did a full migration to ember-glimmer and removed support for angle bracket components I am going to close this PR.

I am sure we will work together on this again when we get angle bracket components going again...
